### ISSUE TITLE
Do not hardcode `<pkg>/jsx-runtime` as entry point

### ISSUE THREAD

**Issue by jvanbruegge** [babel/babel#13069]
## Feature Request

- [x] I would like to work on this feature!

**Is your feature request related to a problem?**
At the moment, as a library author, it is not possible to use the nice automatic import feature of jsx-runtimes without having a second entrypoint because `jsxImportSource` always appends `/jsx-runtime` to the import for no apparent reason.

**Describe the solution you'd like**
Just take the path that is configured by `jsxImportSource` as is so it would also work for the main entry point of a package

**Describe alternatives you've considered**.
There are not really any, I can't get babel to not do the appending

**Documentation, Adoption, Migration Strategy**
This would be a breaking change, so would have to be in babel 8

**Comment by nicolo-ribaudo** [comment]
There isn't a single entry point (`jsx-runtime`), but two of them: `jsx-runtime` and `jsx-dev-runtime`. This allows libraries to provide an optimized production version (`jsx-runtime`) and a developement version with better errors and debugging support.

Also, these paths are part of the API of React-like JSX libraries (which are what `@babel/plugin-transform-react-jsx` supports): any API change should first be discussed with the React team (either in https://github.com/reactjs/rfcs or https://github.com/facebook/react).

I can suggest two alternatives:
- You can use a custom Babel plugin that removes `jsx-runtime` and `jsx-dev-runtime` from your imports:https://astexplorer.net/#/gist/aa534e733598f631ddabd87564a36b2c/latest
- The JSX library you use can point to the main entrypoint for those two paths, adding something this to `exports` in `package.json`:
  ```json
    "./jsx-runtime": "./lib/index.js",
    "./jsx-dev-runtime": "./lib/index.js"
  ```

**Comment by jvanbruegge** [comment]
The `exports` field is not supported by typescript at the moment, so that is not an option. But I can think of something simpler:
Add two new options to the jsx-transform plugin, `jsxRuntime` that defaults to `./jsx-runtime` and `jsxDevRuntime` that defaults to `./jsx-dev-runtime`. That way I could simply use the empty string for both and get the behavior I want and it is not a breaking change any more

**Comment by nicolo-ribaudo** [comment]
Wouldn't something like https://astexplorer.net/#/gist/aa534e733598f631ddabd87564a36b2c/latest work the same then?

Most of the JSX frameworks that rely on `@babel/plugin-transform-react-jsx` already implemented support for the new runtime, so adding two options to customize their names seems a niche use case to me :sweat_smile: (and that's where custom plugins shine, since they give you unlimited customizability)

**Comment by jvanbruegge** [comment]
> we would also need to make sure that that option is accepted by TypeScript

That's actually the reason I am opening this issue here. I am using Typescript and they follow babel in this regard. This is also why a custom plugin won't help me, because I will get a compiler error from Typescript

**Comment by JLHwung** [comment]
> The exports field is not supported by typescript at the moment

Yeah, the TypeScript team is aware of this issue and track it in https://github.com/microsoft/TypeScript/issues/33079.

Before TypeScript supports this feature, you can add two proxy files to your library:

jsx-runtime.js
```js
module.exports = require(".");
```

ditto for `./jsx-dev-runtime`.

You can migrate to exports map when TypeScript eventually implements it.

**Comment by nicolo-ribaudo** [comment]
I haven't checked, but I'd also try _already using `"exports"`_ and just creating `jsx-runtime.d.ts` and `jsx-dev-runtime.d.ts` specifically for TS.
